controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,259 @@
+	}, gatekeeper)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Error(err, "Gatekeeper resource is not found")
+
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Get config or create if not exist
+	config := &v1alpha1.Config{}
+	err = r.Get(ctx, types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      "config",
+	}, config)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			config = &v1alpha1.Config{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "config",
+					Namespace: r.Namespace,
+				},
+			}
+
+			createErr := r.Create(ctx, config)
+			if createErr != nil {
+				log.Error(err, "Fail to create the Gatekeeper Config object, will retry.")
+
+				return reconcile.Result{}, createErr
+			}
+
+			log.Info("The Gatekeeper Config object was created")
+		} else {
+			return reconcile.Result{}, err
+		}
+	}
+
+	constraintPodStatus := &v1beta1.ConstraintPodStatus{}
+
+	err = r.Get(ctx, request.NamespacedName, constraintPodStatus)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Cannot find the ConstraintPodStatus")
+
+			err = r.handleDeleteEvent(ctx, request.Name, config)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{}, nil
+		}
+		// Requeue
+		return reconcile.Result{}, err
+	}
+
+	constraint, constraintName, err := getConstraint(ctx, *constraintPodStatus, r.DynamicClient)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("The Constraint was not found", "constraintName:", constraintName)
+
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	constraintMatchKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+	if err != nil {
+		r.Log.V(1).Info("There are no provided kinds in the Constraint", "constraintName:", constraintName)
+
+		err = r.handleDeleteEvent(ctx, request.Name, config)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil //nolint:nilerr
+	}
+
+	constraintSyncOnlyEntries, err := r.DiscoveryStorage.getSyncOnlys(constraintMatchKinds)
+	if err != nil {
+		if errors.Is(err, ErrNotFoundDiscovery) {
+			r.Log.V(1).Info("Cannot find matched discovery. Requeue after 10 secs")
+
+			requeueTime = time.Second * 10
+		} else {
+			log.Error(err, "Error to get matching kind and apigroup")
+
+			return reconcile.Result{}, err
+		}
+	}
+
+	r.ConstraintToSyncOnly[request.Name] = constraintSyncOnlyEntries
+
+	uniqSyncOnly := r.getUniqSyncOnly()
+
+	if reflect.DeepEqual(uniqSyncOnly, config.Spec.Sync.SyncOnly) {
+		r.Log.V(1).Info("There are no changes detected. Cancel Updating")
+
+		return reconcile.Result{RequeueAfter: requeueTime}, nil
+	}
+
+	config.Spec.Sync.SyncOnly = uniqSyncOnly
+
+	err = r.Update(ctx, config, &client.UpdateOptions{})
+	if err != nil {
+		log.Error(err, "unable to update config syncOnly")
+
+		return reconcile.Result{RequeueAfter: requeueTime}, nil
```

Comment:
You should retry the request if the update fails. It could come from a conflict (i.e. a user updates the Config while the Reconcile was running).
```suggestion
		return reconcile.Result{}, err
```